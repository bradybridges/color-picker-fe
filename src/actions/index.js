export const addProjects = (projects) => ({
  type: 'ADD_PROJECTS',
  projects,
});

export const addPalettes = (palettes) => ({
  type: 'ADD_PALETTES',
  palettes,
});
