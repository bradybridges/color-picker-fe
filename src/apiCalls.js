export const getProjects = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/projects');
  if(!response.ok) {
    return Error('Failed to fetch projects');
  }
  const projects = await response.json();
  return projects;
};

export const getPalettes = async () => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes');
  if(!response.ok) {
    return Error('Failed to fetch palettes');
  }
  const palettes = await response.json();
  return palettes;
};

export const getProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`);
  if(!response.ok) {
    return Error('Failed to fetch pallete');
  }
  const palette = await response.json();
  return palette;
};

export const getPalette = async (paletteId) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${paletteId}`);
  if (!response.ok) {
    return Error('Failed to fetch palette');
  }
  const palette = await response.json();
  return palette;
};

export const postProject = async (name) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  });
  if (!response.ok) {
    return Error('Failed to post new project');
  }
  const message = await response.json();
  console.log('message', message)
  return message;
};

export const postPalette = async (project_id, palette_name, color_1, color_2, color_3, color_4, color_5) => {
  const response = await fetch('https://color-picker-backend.herokuapp.com/api/v1/palettes', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(
      {
        project_id,
        palette_name,
        color_1,
        color_2,
        color_3,
        color_4,
        color_5,
      },
    ),
  });
  if (!response.ok) {
    return response;
  }
  const message = await response.json();
  return message;
};

export const deleteProject = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/projects/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete project');
  }
  const message = await response.json();
  return message;
};

export const deletePalette = async (id) => {
  const response = await fetch(`https://color-picker-backend.herokuapp.com/api/v1/palettes/${id}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    return Error('Failed to delete palette');
  }
  const message = await response.json();
  return message;
};

export const patchProject = async () => {

};

export const patchPalette = async () => {

};
