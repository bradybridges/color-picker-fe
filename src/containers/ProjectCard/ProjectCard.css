.project-card {
  color: white;
  display: flex;
  flex-direction: column;
  height: 250px;
  overflow: scroll;
  width: 100%;
}

