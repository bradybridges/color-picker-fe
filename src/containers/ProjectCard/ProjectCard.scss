.project-card {
  align-items: center;
  color: white;
  display: flex;
  flex-direction: column;
  height: 250px;
  overflow: scroll;
  width: 100%;
}

.delete-modal {
  position: absolute;
  top: 50%;
  left: 35%;
  background-color: #000;
  border: solid #fff 1px;
  border-radius: 4px;
  width: 25%;
  text-align: center;
  padding: 10px 0px;
}

.project-button {
  width: 35%;
  height: 40px;
  border: solid #2d9fcc 1px;
  border-radius: 4px;
  background-color: black;
  color: #fff;
  &:hover {
    box-shadow: 3px -3px 6px 0px #2d9fcc;
    cursor: pointer;
  }
}

.delete-confirmation-container {
  display: flex;
  justify-content: space-around;
}

.project-title-container {
  align-items: center;
  display: flex;
  justify-content: space-evenly;
  width: 100%;
}