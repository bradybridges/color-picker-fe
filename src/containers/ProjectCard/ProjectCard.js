import React, { Component } from 'react';
import { connect } from 'react-redux';
import './ProjectCard.scss';
import Palette from '../Palette/Palette';
import { deleteProject } from '../../apiCalls';
import * as actions from '../../actions';

export class ProjectCard extends Component {
  constructor() {
    super();
    this.state = {
      showDeleteModal: false,
    };
  }
  renderPalettes = () => {
    const { palettes, id } = this.props;
    const currentPalettes = palettes.filter((palette) => palette.project_id === id);
    return currentPalettes.map((palette, index) => {
      return (
        <Palette
          {...palette}
          key={`${palette.id}${index}`}
          className='palette-card'
        />
      );
    })
  }
  renderDeleteModal = () => {
    const { name } = this.props;
    return (
      <div className="delete-modal">
        <h4>Are you want to delete {name}?</h4>
        <div className="delete-confirmation-container">
          <button className="project-button" type="button" onClick={this.handleDeleteProject}>Delete</button>
          <button className="project-button" type="button" onClick={this.handleCancelDelete}>Cancel</button>
        </div>
      </div>
    );
  }
  handleCancelDelete = () => {
    this.setState({ showDeleteModal: false });
  }
  handleDeleteProject = (e) => {
    e.preventDefault();
    const { id, projects, palettes } = this.props;
    deleteProject(id);
    this.updateProjects(id, projects);
    this.updatePalettes(id, palettes);
  }
  showModal = () => {
    this.setState({ showDeleteModal: true });
  }
  updateProjects = (id, projects) => {
    const updatedProjects = projects.filter((project) => project.id !== id);
    this.props.updateProjects(updatedProjects);
  }
  updatePalettes = (id, palettes) => {
    const updatedPalettes = palettes.filter((palette) => palette.project_id !== id);
    this.props.updatePalettes(updatedPalettes);
  }
  render() {
    const { name } = this.props;
    return (
      <section className='project-card'>
        <div className="project-title-container">
          <h4 className='project-name'>{name}</h4>
          <button
            className="project-button"
            type="button"
            onClick={this.showModal}
          >
            Delete Project
          </button>
        </div>
        {this.renderPalettes()}
        {this.state.showDeleteModal && this.renderDeleteModal()}
      </section>
    );
  }
}

export const mapState = (state) => ({
  palettes: state.palettes
});

export const mapDispatch = (dispatch) => ({
  updateProjects: (projects) => dispatch(actions.addProjects(projects)),
  updatePalettes: (palettes) => dispatch(actions.addPalettes(palettes)),
});

export default connect(mapState, mapDispatch)(ProjectCard);
