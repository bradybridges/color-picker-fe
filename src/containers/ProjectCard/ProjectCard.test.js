import React from 'react'
import { shallow } from 'enzyme'
import {ProjectCard, mapState, mapDispatch} from './ProjectCard'
import { deleteProject } from '../../apiCalls'
import { addProjects, addPalettes } from '../../actions'

jest.mock('../../apiCalls')

describe('ProjectCard', () => {
  describe('ProjectCard component', () => {
    let wrapper;
    beforeEach(() => {
      deleteProject.mockImplementation(() => {
        return Promise.resolve({message: 'Successfully deleted project'})
      })
      const mockPalettes = [
          {
            id: 1,
            project_id: 1,
            name: 'a palette',
            color_1: '#123456',
            color_2: '#123456',
            color_3: '#123456',
            color_4: '#123456',
            color_5: '#123456',
          }, 
          {
            id: 2,
            project_id: 1,
            name: 'another palette',
            color_1: '#AABBCC',
            color_2: '#AABBCC',
            color_3: '#AABBCC',
            color_4: '#AABBCC',
            color_5: '#AABBCC',
          }
        ]
      

      const mockProjects = [
        {
          id: 1,
          name: 'a project'
        },
        {
          id: 2,
          name: 'another project'
        }
      ]

      const mockUpdateProjects = jest.fn();
      const mockUpdatePalettes = jest.fn()
      wrapper = shallow(<ProjectCard
        name="Great Project"
        palettes={mockPalettes}
        projects={mockProjects}
        id={1} 
        updateProjects={mockUpdateProjects} 
        updatePalettes={mockUpdatePalettes} 
      />)
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should update showDeleteModal to false in state when handleCancelDelete is invoked', () => {
      // const expected = { showDeleteModal: false }

      wrapper.instance().setState({ showDeleteModal: true})

      wrapper.instance().handleCancelDelete()

      expect(wrapper.state('showDeleteModal')).toEqual(false)
    })

    it('should update showDeleteModal to true in state when showModal is called', () => {
      wrapper.instance().setState({ showDeleteModal: true })

      wrapper.instance().showModal()

      expect(wrapper.state('showDeleteModal')).toEqual(true)
    })

    it('call handleDeleteProject with the correct information', () => {
      const mockEvent = { preventDefault: jest.fn() }
      
      wrapper.instance().updateProjects = jest.fn()
      wrapper.instance().updatePalettes = jest.fn()

      wrapper.instance().handleDeleteProject(mockEvent)

      expect(deleteProject).toHaveBeenCalledWith(1);

      expect(wrapper.instance().updateProjects).toHaveBeenCalledWith(1, [{"id": 1, "name": "a project"}, {"id": 2, "name": "another project"}])

      expect(wrapper.instance().updatePalettes).toHaveBeenCalledWith(1, [
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#123456',
          color_2: '#123456',
          color_3: '#123456',
          color_4: '#123456',
          color_5: '#123456',
        }, 
        {
          id: 2,
          project_id: 1,
          name: 'another palette',
          color_1: '#AABBCC',
          color_2: '#AABBCC',
          color_3: '#AABBCC',
          color_4: '#AABBCC',
          color_5: '#AABBCC',
        }
      ])
      


    })

    it('should call updateProjects with the correct information', () => {
      const result = wrapper.instance().updateProjects(1,[{"id": 1, "name": "a project"}, {"id": 2, "name": "another project"}])

      expect(wrapper.instance().props.updateProjects).toHaveBeenCalledWith([{"id": 2, "name": "another project"}])


    })

    it('should call updatePalettes with the correct information', () => {
      const result = wrapper.instance().updatePalettes(1,[
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#123456',
          color_2: '#123456',
          color_3: '#123456',
          color_4: '#123456',
          color_5: '#123456',
        }, 
        {
          id: 2,
          project_id: 1,
          name: 'another palette',
          color_1: '#AABBCC',
          color_2: '#AABBCC',
          color_3: '#AABBCC',
          color_4: '#AABBCC',
          color_5: '#AABBCC',
        }
      ])

      expect(wrapper.instance().props.updatePalettes).toHaveBeenCalledWith([])


    })
  })

  describe('mapState', () => {
    it('should return an object with the projects array', () => {
      const mockState = {
        projects: [
          {
            id: 1,
            name: 'a project'
          }, 
          {
            id: 2,
            name: 'another project'
          }
        ],
        palettes: [
          {
            id: 1,
            project_id: 1,
            name: 'a palette',
            color_1: '#FFFFFF',
            color_2: '#FFFFFF',
            color_3: '#FFFFFF',
            color_4: '#FFFFFF',
            color_5: '#FFFFFF',
          }
        ]
      }

      const expectedState = {
        projects: [
          {
            id: 1,
            name: 'a project'
          }, 
          {
            id: 2,
            name: 'another project'
          }
        ],
        palettes: [
          {
            id: 1,
            project_id: 1,
            name: 'a palette',
            color_1: '#FFFFFF',
            color_2: '#FFFFFF',
            color_3: '#FFFFFF',
            color_4: '#FFFFFF',
            color_5: '#FFFFFF',
          }
        ]
      }

      const mappedProps = mapState(mockState)

      expect(mappedProps).toEqual(expectedState)
    })
  })

  describe('mapDispatch', () => {
    it('calls dispatch with a projects array when updateProjects is called', () => {
      const mockDispatch = jest.fn()

      const actionToDispatch = addProjects([{id: 1, name: 'a name'}, {id: 2, name: 'another name'}])

      const mappedProps = mapDispatch(mockDispatch)

      mappedProps.updateProjects([{id: 1, name: 'a name'}, {id: 2, name: 'another name'}])

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('calls dispatch with a palettes array when updatePalettes is called', () => {
      const mockDispatch = jest.fn()

      const actionToDispatch = addPalettes([
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#FFFFFF',
          color_2: '#FFFFFF',
          color_3: '#FFFFFF',
          color_4: '#FFFFFF',
          color_5: '#FFFFFF',
        }
      ])

      const mappedProps = mapDispatch(mockDispatch)

      mappedProps.updatePalettes([
        {
          id: 1,
          project_id: 1,
          name: 'a palette',
          color_1: '#FFFFFF',
          color_2: '#FFFFFF',
          color_3: '#FFFFFF',
          color_4: '#FFFFFF',
          color_5: '#FFFFFF',
        }
      ])

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})
