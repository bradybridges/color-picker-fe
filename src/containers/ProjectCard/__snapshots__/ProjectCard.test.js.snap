// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectCard ProjectCard component should match the snapshot 1`] = `
<section
  className="project-card"
>
  <div
    className="project-title-container"
  >
    <h4
      className="project-name"
    >
      Great Project
    </h4>
    <button
      className="project-button"
      onClick={[Function]}
      type="button"
    >
      Delete Project
    </button>
  </div>
  <Connect(Palette)
    className="palette-card"
    color_1="#123456"
    color_2="#123456"
    color_3="#123456"
    color_4="#123456"
    color_5="#123456"
    id={1}
    key="10"
    name="a palette"
    project_id={1}
  />
  <Connect(Palette)
    className="palette-card"
    color_1="#AABBCC"
    color_2="#AABBCC"
    color_3="#AABBCC"
    color_4="#AABBCC"
    color_5="#AABBCC"
    id={2}
    key="21"
    name="another palette"
    project_id={1}
  />
</section>
`;
